cng/hkdf.go

```diff
@@ -37,45 +37,77 @@ type hkdf struct {
 	info []byte
 
 	hashLen int
-	buf     []byte
+	n       int // count of bytes requested from Read
+	// buf contains the derived data.
+	// len(buf) can be larger than n, as Read may derive
+	// more data than requested and cache it in buf.
+	buf []byte
 }
 
 func (c *hkdf) finalize() {
 	bcrypt.DestroyKey(c.hkey)
 }
 
-func (c *hkdf) Read(p []byte) (int, error) {
+func hkdfDerive(hkey bcrypt.KEY_HANDLE, info, out []byte) (int, error) {
 	var params *bcrypt.BufferDesc
-	if len(c.info) > 0 {
+	if len(info) > 0 {
 		params = &bcrypt.BufferDesc{
 			Count: 1,
 			Buffers: &bcrypt.Buffer{
-				Length: uint32(len(c.info)),
+				Length: uint32(len(info)),
 				Type:   bcrypt.KDF_HKDF_INFO,
-				Data:   uintptr(unsafe.Pointer(&c.info[0])),
+				Data:   uintptr(unsafe.Pointer(&info[0])),
 			},
 		}
+		defer runtime.KeepAlive(params)
 	}
+	var n uint32
+	err := bcrypt.KeyDerivation(hkey, params, out, &n, 0)
+	return int(n), err
+}
+
+func (c *hkdf) Read(p []byte) (int, error) {
 	// KeyDerivation doesn't support incremental output, each call
 	// derives the key from scratch and returns the requested bytes.
 	// To implement io.Reader, we need to ask for len(c.buf) + len(p)
 	// bytes and copy the last derived len(p) bytes to p.
-	// We use c.buf to know how many bytes we've already derived and
-	// to avoid allocating the whole output buffer on each call.
-	prevLen := len(c.buf)
-	needLen := len(p)
-	remains := 255*c.hashLen - prevLen
-	// Check whether enough data can be generated.
-	if remains < needLen {
+	maxDerived := 255 * c.hashLen
+	totalDerived := c.n + len(p)
+	// Check whether enough data can be derived.
+	if totalDerived > maxDerived {
 		return 0, errors.New("hkdf: entropy limit reached")
 	}
-	c.buf = append(c.buf, make([]byte, needLen)...)
-	var size uint32
-	if err := bcrypt.KeyDerivation(c.hkey, params, c.buf, &size, 0); err != nil {
-		return 0, err
+	// Check whether c.buf already contains enough derived data,
+	// otherwise derive more data.
+	if bytesNeeded := totalDerived - len(c.buf); bytesNeeded > 0 {
+		// It is common to derive multiple equally sized keys from the same HKDF instance.
+		// Optimize this case by allocating a buffer large enough to hold
+		// at least 3 of such keys each time there is not enough data.
+		blocks := bytesNeeded / c.hashLen
+		if bytesNeeded%c.hashLen != 0 {
+			// Round up to the next multiple of hashLen.
+			blocks += 1
+		}
```

Comment:
I remember using this in one of our repos, but I forget which:
```suggestion
		// Round up to the next multiple of hashLen.
		blocks := (bytesNeeded-1)/c.hashLen + 1
```